### PR TITLE
First commit of BPM to ms converter Goodie

### PR DESCRIPTION
**Pull request by stefolof**
# Goodie Pull Request Template

Using this template will help us better understand your Instant Answer and assist you when necessary.  Simply copy and paste the markdown below into the description of your GitHub pull request and complete as appropriate.

**What does your Instant Answer do?**
It converts BPM (beats per minute) into note values in milliseconds. It is useful in musical applications for example when needing to calculate delay times, reverb times, compressor attack and release settings etc.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
There exists sites with bmp to ms calculators but it is more convenient to get the results directly in the search engine.

**What is the data source for your Instant Answer? (Provide a link if possible)**
The calculation is done inside the Perl script.

**What are some example queries that trigger this Instant Answer?**
120 bpm to ms
90 bpm to ms

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Music performers and recording communities.

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No

**Are you having any problems? Do you need our help with anything?**
I did not seem to be able to write a test file for the html, only the plain-text answer.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
bpmtoms_example

## Checklist
Please place an 'X' where appropriate.

```
[x] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
[x] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
[x] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[x] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[x] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [x] Google Chrome
        [x] Firefox
        [] IE 7
        [x] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [x] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

### CODE REVIEW

**Comment by abeyang** [comment]
Hi @stefolof, while we're still working on some things internally, I don't want you to hang around in limbo. In the meantime, what do you think about implementing this design:

bpm-to-ms-grid-v1 1

We can use the [whois template](https://duckduckgo.com/?q=whois+apple.com&ia=whois), along with these additional changes: https://github.com/duckduckgo/zeroclickinfo-spice/issues/1414

That way, we can push this live, and revisit once all the other hooks are in place. LMK if you have any questions!

**Comment by stefolof** [comment]
Hi @abeyang, sounds good to me! I'll look into it.

**Comment by stefolof** [comment]
One quick, are the CSS changes in https://github.com/duckduckgo/zeroclickinfo-spice/issues/1414 merge or replace?

**Comment by jagtalon** [comment]
That looks cool, @abeyang! @stefolof replace if it already exists in https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/whois/whois.css. As long as it gets applied, I guess. :)

Using CSS in Goodies is kinda clunky at the moment. You'd have to read in the file (https://duck.co/duckduckhack/goodie_advanced_handle_functions#using-files-in-the-share-directory) and then concatenate that into the HTML output.

**Comment by mintsoft** [comment]
@jagtalon css is automatically included if it has the same name in Goodies! i.e. /share/subnet/subnet.css

**Comment by stefolof** [comment]
I'm thinking we could get rid of the "ms" after each number as it says milliseconds in the headline. What do you think?

bpmtoms_screen

**Comment by abeyang** [comment]
@stefolof Good point -- I agree! I know you're still working on it, but can you right-align the numbers (so the zeroes should line up)?

**Comment by stefolof** [comment]
bpmtoms_screen2

**Comment by abeyang** [comment]
Looking nice. Can we decrease the gutter between key and value? So the gap should decrease between whole note/1920; triplet/1280; dotted/2880

**Comment by stefolof** [comment]
Sure, I'll be back tomorrow.

**Comment by stefolof** [comment]
Current look:
bpmtoms_screen3

**Comment by mintsoft** [comment]
@abeyang that look good to you?

**Comment by jagtalon** [comment]
Putting it up on a server!

**Comment by jagtalon** [comment]
@abeyang https://ddh1.duckduckgo.com/?q=125+bpm+to+ms&ia=answer

File: lib/DDG/Goodie/BPMToMs.pm [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    my $dotted_whole_note = 360000;
+    
+    my @divisors = map { 2 ** $_ } 0 .. 5; # Create a list of divisors to calculate the values of half notes, quarter notes etc.
+    
+    my @straight_values = map { int( $straight_whole_note / ($bpm * $_) + 0.5) } @divisors;
+    my @triplet_values = map { int( $triplet_whole_note / ($bpm * $_) + 0.5) } @divisors;
+    my @dotted_values = map { int( $dotted_whole_note / ($bpm * $_) + 0.5) } @divisors;
+    
+    my $html_content = "<div class=\"bpmto_ms\">
+                        <h3 class=\"zci__header\">$bpm bpm in milliseconds</h3>
```
Comment (jagtalon) [inline lib/DDG/Goodie/BPMToMs.pm]: These should probably encoded like in https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Uppercase.pm#L35 CC @mintsoft

Comment (jagtalon) [inline lib/DDG/Goodie/BPMToMs.pm]: Just to prevent some XSS stuff

Comment (mintsoft) [inline lib/DDG/Goodie/BPMToMs.pm]: Absolutely; there's no validation on this being an integer at all

**Comment by stefolof** [comment]
Thanks, guys!

As this Goodie preferably returns html, should the tests be written for html or for plain-text (or both)? Can I omit the plain-text version altogether?

**Comment by abeyang** [comment]
@stefolof it's looking much better! Can you make these css adjustments, though:

```css
.bpmto_ms .record > td {
  border-bottom: 0;
}

.zci.is-active {
  padding-bottom: 0.5em;
}
```

(Or, for the first one, remove border-bottom altogether from the css.)

Thanks -- almost there!

**Comment by mintsoft** [comment]
@stefolof Ideally both; currently people tend to test for accuracy using the text output but adding some key tests with regex around the html output: 
https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/t/HTMLEntitiesEncode.t#L74
https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/t/CalendarToday.t#L92

**Comment by stefolof** [comment]
Not sure I'm going about this the right way. When wrapping the result inside a structured_answer, the font gets bigger.

bpmtoms_screen4

**Comment by jagtalon** [comment]
@stefolof maybe we should just skip the structured_answer and just do `return html => ...` like before

**Comment by stefolof** [comment]
@jagtalon Yes, it seems to me when looking through the Goodies using structured_answer that they all give more of one-line type answer - not html tables. If anyone else has a suggestion, please chime in.

**Comment by jagtalon** [comment]
@abeyang Here's the latest one! https://ddh1.duckduckgo.com/?q=120+bpm+to+ms&ia=answer

@abeyang @stefolof Unfortunately we can't have `.zci.is-active` because it's too global and affects other IAs.

**Comment by abeyang** [comment]
@stefolof we still need to have the css tweaks as mentioned in https://github.com/duckduckgo/zeroclickinfo-goodies/pull/803#issuecomment-70723133. Basically, want to remove that last border-bottom and reduce the whitespace below it.

**Comment by stefolof** [comment]
@abeyang, the border-bottom should be fixed by 15ee104.

**Comment by abeyang** [comment]
@stefolof doh! I was looking at a cached version. My bad.

However, it seems like the padding-bottom is being overridden by style.css:

pasted_image_1_23_15__10_13_am

Any way we can force it to be 0.5em (instead of the default 1em)?

**Comment by stefolof** [comment]
@abeyang, no worries. :)  Regarding `.zci.is-active`, I think @jagtalon can answer that better than me.

**Comment by jagtalon** [comment]
@stefolof I guess you could use `.zci.zci--answer` instead of `.zci.is-active`

**Comment by stefolof** [comment]
Maybe someone can help me understand why the Travis build is failing sometimes. Yesterday it worked fine, the day before that it failed.

**Comment by moollaza** [comment]
@stefolof sorry about the Travis builds, they can be finnicky at times. Looking at them, they're not failing because of your code, there was a bug that was since fixed that had caused a lot of them to fail. I'm re-running the build on your latest commit to make sure it works, but I expect it should have no problems.

**Comment by stefolof** [comment]
@moollaza, thanks for the update! I guess we're all set then?

**Comment by moollaza** [comment]
@stefolof this is *juuust* about done but I noticed a couple things:

1. It's broken on mobile devices.  All the text compresses horizontally and overlaps:

    screen shot 2015-02-03 at 21 23 22

2. We're using empty `<td>` for padding -- why not CSS?

**Comment by stefolof** [comment]
Good catch, @moollaza. I may have to rethink the layout for mobile. Seems like the rows will be too long.

**Comment by abeyang** [comment]
@stefolof would you be ok if, for this version, we just lose the triplet and dotted columns altogether? We're still hoping to get in the tiled-version up really soon, and that may "fix" the mobile view on its own. :smile:

**Comment by stefolof** [comment]
@abeyang yeah, I think that's OK. 90% of the time people will want the straight values. And it seems I'd have to break the table structure to accommodate all that text on a mobile device. It will be better when we can use symbols.

**Comment by moollaza** [comment]
@abeyang that's a great suggestion. If you're okay with it, @stefolof I think that's the easiest way to get this IA ready for merge.

**Comment by stefolof** [comment]
@moollaza yes, it's the best solution, I'll try and have it ready tomorrow

**Comment by moollaza** [comment]
> @moollaza yes, it's the best solution, I'll try and have it ready tomorrow

Wonderful, thanks again! And just so you know, there's absolutely no rush :)

**Comment by stefolof** [comment]
Great, because I didn't manage today. :) Soon though ... Thanks guys for all the help.

**Comment by mintsoft** [comment]
@stefolof No pressure :)

**Comment by stefolof** [comment]
Here's how it looks without triplet and dotted:

ddg_simple_screen

Also checked on Safari / iPhone 4 and it looked alright, no overlapping text.

**Comment by moollaza** [comment]
@stefolof thanks a lot! Looks great. @abeyang thoughts?

**Comment by abeyang** [comment]
Looks good! :+1:

**Comment by mintsoft** [comment]
:+1:

**Comment by moollaza** [comment]
haha @mintsoft you beat me to it!

I was just about to say that the bottom border on the last row is back @stefolof 
I'll fix it in a new PR

**Comment by moollaza** [comment]
Fixed in #989

**Comment by stefolof** [comment]
Good catch again, @moollaza! I missed the bottom-border on mobile. Great job, guys. Is there a thread I can watch for updates to the design framework?

**Comment by moollaza** [comment]
@stefolof if you mean updated on the Goodie templates, I would suggest you subscribe to our developer email list: duckduckhack@listbox.com (usually 1 or 2 emails a week), we'll be sure to put updates there. Also you could watch our Documentation Repo as it will be updated once everything is ready.

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & issues or even submit more Instant Answers!

**Comment by moollaza** [comment]
@stefolof congrats! It's live: https://duckduckgo.com/?q=120+bpm+to+ms&ia=answer

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Comment by mintsoft** [comment]
Awesome Congrats @stefolof

**Comment by stefolof** [comment]
Thanks very much, guys! That's super cool! :)

I have another idea for an IA. Just a bit snowed under with work at the moment.

**Comment by mintsoft** [comment]
@stefolof you can always put it as an idea on https://duck.co/ideas Maybe someone will implement it for you!

**Pull request merged**